### PR TITLE
Slow operations are prohibited on EDT

### LINKED ISSUE #219: Slow operations are prohibited on EDT
**Describe the bug**
Slow operations are prohibited on EDT error message during the interations with local project files.

**To Reproduce**
1. Generate tests

**Expected behavior**
No such errors.

**Additional context**
Error message:
```
java.lang.Throwable: Slow operations are prohibited on EDT. See SlowOperations.assertSlowOperationsAreAllowed javadoc.
	at com.intellij.openapi.diagnostic.Logger.error(Logger.java:376)
	at com.intellij.util.SlowOperations.assertSlowOperationsAreAllowed(SlowOperations.java:106)
	at com.intellij.workspaceModel.core.fileIndex.impl.WorkspaceFileIndexDataImpl.ensureIsUpToDate(WorkspaceFileIndexDataImpl.kt:142)
	at com.intellij.workspaceModel.core.fileIndex.impl.WorkspaceFileIndexDataImpl.getFileInfo(WorkspaceFileIndexDataImpl.kt:87)
	at com.intellij.workspaceModel.core.fileIndex.impl.WorkspaceFileIndexImpl.getFileInfo(WorkspaceFileIndexImpl.kt:245)
	at com.intellij.workspaceModel.core.fileIndex.impl.WorkspaceFileIndexImpl.findFileSet(WorkspaceFileIndexImpl.kt:201)
	at com.intellij.openapi.roots.impl.ProjectFileIndexImpl.getClassRootForFile(ProjectFileIndexImpl.java:113)
	at com.intellij.psi.impl.file.impl.JavaFileManagerImpl.hasAcceptablePackage(JavaFileManagerImpl.java:125)
	at com.intellij.psi.impl.file.impl.JavaFileManagerImpl.doFindClasses(JavaFileManagerImpl.java:92)
	at com.intellij.psi.impl.file.impl.JavaFileManagerImpl.findClass(JavaFileManagerImpl.java:105)
	at com.intellij.psi.impl.PsiElementFinderImpl.findClass(PsiElementFinderImpl.java:40)
	at com.intellij.psi.impl.JavaPsiFacadeImpl.doFindClass(JavaPsiFacadeImpl.java:95)
	at com.intellij.psi.impl.JavaPsiFacadeImpl.findClass(JavaPsiFacadeImpl.java:73)
	at com.intellij.psi.impl.PsiClassImplUtil.findSpecialSuperClass(PsiClassImplUtil.java:740)
	at com.intellij.psi.impl.PsiClassImplUtil.getSuperClass(PsiClassImplUtil.java:733)
	at com.intellij.psi.impl.source.PsiClassImpl.getSuperClass(PsiClassImpl.java:207)
	at org.jetbrains.research.testspark.tools.llm.Llm.collectClassesToTest(Llm.kt:69)
	at org.jetbrains.research.testspark.tools.llm.Llm.getLLMProcessManager$lambda$0(Llm.kt:40)
	at com.intellij.openapi.application.impl.RwLockHolder.runReadAction(RwLockHolder.kt:271)
	at com.intellij.openapi.application.impl.ApplicationImpl.runReadAction(ApplicationImpl.java:845)
	at org.jetbrains.research.testspark.tools.llm.Llm.getLLMProcessManager(Llm.kt:38)
	at org.jetbrains.research.testspark.tools.llm.Llm.generateTestsForClass(Llm.kt:94)
	at org.jetbrains.research.testspark.actions.TestSparkAction$TestSparkActionWindow.startLLMGeneration(TestSparkAction.kt:359)
	at org.jetbrains.research.testspark.actions.TestSparkAction$TestSparkActionWindow.addListeners$lambda$9(TestSparkAction.kt:308)
	at java.desktop/javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:1972)
	at java.desktop/javax.swing.AbstractButton$Handler.actionPerformed(AbstractButton.java:2313)
	at java.desktop/javax.swing.DefaultButtonModel.fireActionPerformed(DefaultButtonModel.java:405)
	at java.desktop/javax.swing.DefaultButtonModel.setPressed(DefaultButtonModel.java:262)
	at java.desktop/javax.swing.plaf.basic.BasicButtonListener.mouseReleased(BasicButtonListener.java:279)
	at java.desktop/java.awt.Component.processMouseEvent(Component.java:6657)
	at java.desktop/javax.swing.JComponent.processMouseEvent(JComponent.java:3385)
	at java.desktop/java.awt.Component.processEvent(Component.java:6422)
	at java.desktop/java.awt.Container.processEvent(Container.java:2266)
	at java.desktop/java.awt.Component.dispatchEventImpl(Component.java:5027)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2324)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4855)
	at java.desktop/java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4969)
	at java.desktop/java.awt.LightweightDispatcher.processMouseEvent(Container.java:4583)
	at java.desktop/java.awt.LightweightDispatcher.dispatchEvent(Container.java:4524)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2310)
	at java.desktop/java.awt.Window.dispatchEventImpl(Window.java:2809)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4855)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:794)
	at java.desktop/java.awt.EventQueue$3.run(EventQueue.java:739)
	at java.desktop/java.awt.EventQueue$3.run(EventQueue.java:733)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:97)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:766)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:764)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:763)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.kt:698)
	at com.intellij.ide.IdeEventQueue.dispatchMouseEvent(IdeEventQueue.kt:637)
	at com.intellij.ide.IdeEventQueue._dispatchEvent$lambda$10(IdeEventQueue.kt:584)
	at com.intellij.openapi.application.impl.RwLockHolder.runWithEnabledImplicitRead(RwLockHolder.kt:138)
	at com.intellij.openapi.application.impl.RwLockHolder.runWithImplicitRead(RwLockHolder.kt:129)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.kt:584)
	at com.intellij.ide.IdeEventQueue.access$_dispatchEvent(IdeEventQueue.kt:77)
	at com.intellij.ide.IdeEventQueue$dispatchEvent$processEventRunnable$1$1$1.compute(IdeEventQueue.kt:362)
	at com.intellij.ide.IdeEventQueue$dispatchEvent$processEventRunnable$1$1$1.compute(IdeEventQueue.kt:361)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computePrioritized(CoreProgressManager.java:843)
	at com.intellij.ide.IdeEventQueue$dispatchEvent$processEventRunnable$1$1.invoke(IdeEventQueue.kt:361)
	at com.intellij.ide.IdeEventQueue$dispatchEvent$processEventRunnable$1$1.invoke(IdeEventQueue.kt:356)
	at com.intellij.ide.IdeEventQueueKt.performActivity$lambda$1(IdeEventQueue.kt:1021)
	at com.intellij.openapi.application.TransactionGuardImpl.performActivity(TransactionGuardImpl.java:114)
	at com.intellij.ide.IdeEventQueueKt.performActivity(IdeEventQueue.kt:1021)
	at com.intellij.ide.IdeEventQueue.dispatchEvent$lambda$7(IdeEventQueue.kt:356)
	at com.intellij.openapi.application.impl.RwLockHolder.runIntendedWriteActionOnCurrentThread(RwLockHolder.kt:209)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:830)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.kt:398)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:207)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:128)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:117)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:105)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:92)
```

### PR DESCRIPTION
**Pull request by arksap2002**
# Description of changes made
Slow operations are prohibited on EDT error message during the test samples collecting.

# Other notes
Closes #219 

- [x] I have checked that I am merging into correct branch

### CODE REVIEW

**Review by Vladislav0Art — APPROVED** [review]

File: src/main/kotlin/org/jetbrains/research/testspark/tools/llm/Llm.kt [hunk 1]
```diff
@@ -35,11 +36,11 @@ class Llm(override val name: String = "LLM") : Tool {
     private fun getLLMProcessManager(project: Project, psiFile: PsiFile, caretOffset: Int, testSamplesCode: String): LLMProcessManager {
         val classesToTest = mutableListOf<PsiClass>()
 
-        ApplicationManager.getApplication().runReadAction(
-            Computable {
+        ProgressManager.getInstance().runProcessWithProgressSynchronously({
+            ApplicationManager.getApplication().runReadAction {
```
Comment (Vladislav0Art) [inline src/main/kotlin/org/jetbrains/research/testspark/tools/llm/Llm.kt:40]: If it runs synchronously, does it block the main thread during the execution (as it was before for test compilation)?

Comment (Vladislav0Art) [inline src/main/kotlin/org/jetbrains/research/testspark/tools/llm/Llm.kt:40]: If it runs synchronously, does it block the main thread during the execution (as it was before for test compilation)?

**Review by Vladislav0Art — COMMENTED** [review]
If `runProcessWithProgressSynchronously` blocks the user's interaction with the IDE while the operation is in progress, it should be re-implemented in a non-blocking way. A while ago, there was a [thread](https://intellij-support.jetbrains.com/hc/en-us/community/posts/206126669/comments/206720889) about it on the IntelliJ support website.

**Review by Vladislav0Art — COMMENTED** [review]
If `runProcessWithProgressSynchronously` blocks the user's interaction with the IDE while the operation is in progress, it should be re-implemented in a non-blocking way. A while ago, there was a [thread](https://intellij-support.jetbrains.com/hc/en-us/community/posts/206126669/comments/206720889) about it on the IntelliJ support website.

File: src/main/kotlin/org/jetbrains/research/testspark/tools/llm/Llm.kt (hunk 1 above)
Comment (arksap2002) [reply src/main/kotlin/org/jetbrains/research/testspark/tools/llm/Llm.kt:40]: I think, that is the worst case, it wil lead to lackness of test samples. Not blocking the main thread.

Comment (Vladislav0Art) [reply src/main/kotlin/org/jetbrains/research/testspark/tools/llm/Llm.kt:40]: I added the sleep function call to test the blocking:
```kotlin
ApplicationManager.getApplication().runReadAction {
    collectClassesToTest(project, classesToTest, psiFile, caretOffset)
    sleepCancellable(20000)
}
// if sleepCancellable(20000) is placed outside the runReadAction but within the runProcessWithProgressSynchronously, it still blocks
```
It blocks the entire UI; here on the image I cannot click anything within the IDE except for dragging the popup window:

Screenshot 2024-05-24 at 16 32 28

It is generally better to have a non-freezing UI, but I have not yet found a solution to this. Maybe something like `ProgressManager.getInstance().runProcessWithProgressAsynchronously()` may help.

Comment (arksap2002) [reply src/main/kotlin/org/jetbrains/research/testspark/tools/llm/Llm.kt:40]: thank you so much for testing it, yes, that's an incorrect soluiton from my side. I'll work on it

**Comment by arksap2002** [comment]
@Vladislav0Art I agree that it is not logical that we can block this thread in this way. But I use this thread to avoid the `Slow operations are prohibited on EDT` error exclusively. And moreover, I have no need to parallel this piece of code, because I need to use the `classesToTest` array immediately

**Comment by Vladislav0Art** [comment]
> @Vladislav0Art I agree that it is not logical that we can block this thread in this way. But I use this thread to avoid the `Slow operations are prohibited on EDT` error exclusively. And moreover, I have no need to parallel this piece of code, because I need to use the `classesToTest` array immediately

The problem is that if the function call takes a while to be executed, a user cannot do **literally anything** on the UI: no scrolling of the opened file, no file opening, no settings view, nothing. I understand that it fixes the problem with EDT, but the price we pay is somewhat comparable to the problem itself. Unfortunately, at first glance, I see no easy solution that would fix both issues and would not require a lot of refactoring of the execution flow. I think it is a good idea to ask for help from someone from the IntelliJ team; they may suggest something.

If we integrate the UI blocking mechanisms into the codebase, it must be refactored in the future. We just postpone the refactoring, which will make it harder as the codebase grows.

@pderakhshanfar, @NikolaiSviridov, we need your thoughts on that.

**Comment by NikolaiSviridov** [comment]
I feel like it's better to push the current "cosmetic" solution (only solves EDT error without solving the UI freeze) and create a new high-priority task for rewriting logic itself

**Review by Vladislav0Art — APPROVED** [review]
I am following the suggestion of Nikolai. We then need to open an issue with the freezing UI problem.

**Review by NikolaiSviridov — APPROVED** [review]

**Pull request merged**